src/cpu/ppc64/ppc64_gemm_s8x8s32.cpp

```diff
@@ -0,0 +1,185 @@
+/*******************************************************************************
+* Copyright 2022 IBM Corporation
+*   
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#ifdef __MMA__
+#include <altivec.h>
+#include "cpu/ppc64/ppc64_gemm_s8x8s32.hpp"
+#include "cpu/simple_q10n.hpp"
+
+namespace dnnl {
+namespace impl {
+namespace cpu {
+namespace ppc64 {
+
+dnnl_status_t cblas_gemm_s8u8s32_ppc64(int ATflag, int BTflag,
+        char const *offsetc, int m, int n, int k, float alpha,
+        signed char const *A, int lda, signed char const *ao,
+        unsigned char const *B, int ldb, unsigned char const *bo, int *C,
+        float beta, int ldc, int const *co, int flipB_flag) {
+
+    int m_cap, n_cap, k_cap;
+    m_cap = (m + 3) & (~3);
+    n_cap = (n + 3) & (~3);
+    k_cap = (k + 3) & (~3);
+
+    if ((*ao != 0) || (*bo != 0)) {
+        short *Ashort, *AP, *APraw;
+        short *Bshort, *BP, *BPraw;
+        int a_size = lda * (ATflag ? m - 1 : k - 1) + (ATflag ? k : m);
+        int b_size = ldb * (BTflag ? k - 1 : n - 1) + (BTflag ? n : k);
+        Ashort = (short *)malloc(a_size * sizeof(short), 4096);
+        Bshort = (short *)malloc(b_size * sizeof(short), 4096);
+        if (utils::any_null(Ashort, Bshort)) {
+            free(Ashort);
+            free(Bshort);
+            return dnnl_out_of_memory;
+        }
+        for (int i = 0; i < a_size; ++i)
+            Ashort[i] = ((short)A[i]) - (short)*ao;
+        if (flipB_flag) {
+            const signed char *Bflip = (const signed char *)B;
+            const signed char *bo_flip = (const signed char *)bo;
+            for (int i = 0; i < b_size; ++i)
+                Bshort[i] = ((short)(Bflip[i])) - (short)*bo_flip;
+        } else {
+            for (int i = 0; i < b_size; ++i)
+                Bshort[i] = ((short)B[i]) - (short)*bo;
+        }
+        APraw = (short *)malloc((m_cap * k_cap + 15) * sizeof(short), 4096);
+        BPraw = (short *)malloc((k_cap * n_cap + 15) * sizeof(short), 4096);
+        if (utils::any_null(APraw, BPraw)) {
+            free(APraw);
+            free(BPraw);
+            return dnnl_out_of_memory;
```

Comment:
I guess `Ashort` and `Bshort` will leak here.

Does this platform support clang compiler? If yes, did you have a chance to run oneDNN set of gtests named `test_gemm_x8x8s32`, where x is either `s` or `u` with different sanitizers to make sure that code doesn't expose any issues?

Reply:
I do not understand what you mean by "leak".  Memory leak, or some sort of overflow/underflow problem?
Also, I do not know what a "sanitizer" is.

We can compile this code using gcc or clang.

We routinely run all four of the tests you mentioned.
test_gemm_u8u8s32 and test_gemm_s8u8s32 trivially pass because those combinations are not supported.
test_gemm_u8s8s32 and test_gemm_s8s8s32 both pass all 293 tests.

Reply:
`Ashort` and `Bshort` are dynamically allocated. They are validated at the point of allocation and that's it. Now, if `APraw` or `BPraw` fail the allocation, then `Ashort` and `Bshort` will not be freed at this moment, since the section below only frees `APraw` and `BPraw`, but not `Ashort` and `Bshort`. This is a memory leak, basically.

"Sanitizers" are usually referred to special `-fsanitize` instrumentation [options](https://gcc.gnu.org/onlinedocs/gcc/Instrumentation-Options.html). oneDNN supports these options internally through build flags: `-DDNNL_USE_CLANG_SANITIZER=<sanitizer_name>` where `<sanitizer_name>` follows the option value name like "Address" or "Undefined". Once built with these options with clang compiler, runtime will do special instrumentation and might find additional issues which hard to find with manual code review.

Reply:
So, I looked at the page that you pointed us at, and I see ten options for -fsanitize:
   address
   kernel-address
   hwaddress
   kernel-hwaddress
   pointer-compare
   pointer-subtract
   shadow-call-stack
   thread
   leak
   undefined
   
Also, there are 22 sub-options under the "undefined" option.

Question 1: which ones would you like us to test?

Question 2: is it OK if we just use the `-fsanitize=[option]` construction rather than the `-DDNNL_USE_CLANG_SANITIZER` construction, so that we can only apply the build option to our code rather than all of OneDNN?

Reply:
@dzarukin see above...

Reply:
My recommendation to focus on Address one and Undefined one. Those will cover most errors if anything pops up. They also include some sub-options in them.

You should be fine to enable it for the whole project, there shouldn't be any (at least, the we were aware of) issues associated with the piece you are modifying. Thank you.

Reply:
Status update:
First, THANK YOU for your guidance!  Much appreciatedl
As it happens, there was a bug in the Clang compiler for PPC64, but our team has produced a fix which will be eventually promoted into the official distribution.
With this fix, we are able to compile and run all of OneDNN using Clang on our Power10 processor, and it passes all tests.
More importantly, we do see that when we add the sanitizer flag (specifying Address), we find three failures in the "make test" suite, which are clearly in my code.  I'm running the various "make test_benchdnn_*" options now to see how many errors there are there.  My work continues, and I will keep you informed!

Reply:
In all the test_benchdnn_* tests, I only uncovered one failure, when I specified "Address." I found and fixed two bugs and now both test suites run clean. I am going to compile with "Undefined" now.

Reply:
Update: found ONE more bug, hope to have it tested and passing all test suited within 24 to 48 hours.
When is the cutoff?

Reply:
Feel free to take your time to resolve issues. We appreciate higher quality, and will backport to the version needed. Thank you.